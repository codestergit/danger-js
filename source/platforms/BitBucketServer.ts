import { GitJSONDSL, GitDSL } from "../dsl/GitDSL"
import { BitBucketServerPRDSL, BitBucketServerJSONDSL } from "../dsl/BitBucketServerDSL"
import { BitBucketServerAPI } from "./bitbucket_server/BitBucketServerAPI"
import gitDSLForBitBucketServer from "./bitbucket_server/BitBucketServerGit"
import { Platform, Comment } from "./platform"

import * as debug from "debug"

/** Handles conforming to the Platform Interface for BitBucketServer, API work is handle by BitBucketServerAPI */

export class BitBucketServer implements Platform {
  private readonly d = debug("danger:BitBucketServer")
  name: string

  constructor(public readonly api: BitBucketServerAPI) {
    this.name = "BitBucketServer"
  }

  /**
   * Get the Code Review description metadata
   *
   * @returns {Promise<any>} JSON representation
   */
  getReviewInfo = (): Promise<BitBucketServerPRDSL> => this.api.getPullRequestInfo()

  /**
   * Get the Code Review diff representation
   *
   * @returns {Promise<GitDSL>} the git DSL
   */
  getPlatformGitRepresentation = (): Promise<GitJSONDSL> => gitDSLForBitBucketServer(this.api)

  /**
   * Gets inline comments for current PR
   */
  getInlineComments = async (dangerID: string): Promise<Comment[]> => this.api.getDangerInlineComments(dangerID)

  /**
   * Fails the current build, if status setting succeeds
   * then return true.
   */

  updateStatus = async (passed: boolean, message: string, url?: string): Promise<boolean> => {
    const pr = await this.api.getPullRequestInfo()
    const { latestCommit } = pr.fromRef
    try {
      await this.api.postBuildStatus(latestCommit, {
        state: passed ? "SUCCESSFUL" : "FAILED",
        key: "danger.systems",
        name: process.env["PERIL_INTEGRATION_ID"] ? "Peril" : "Danger",
        url: url || "http://danger.systems/js",
        description: message,
      })
      return true
    } catch (error) {
      return false
    }
  }

  /**
   * Returns the `bitBucket_server` object on the Danger DSL
   *
   * @returns {Promise<BitBucketServerDSL>} JSON response of the DSL
   */
  getPlatformDSLRepresentation = async (): Promise<BitBucketServerJSONDSL> => {
    let pr: BitBucketServerPRDSL
    try {
      pr = await this.getReviewInfo()
    } catch {
      process.exitCode = 1
      throw `
        Could not find pull request information,
        perhaps Danger does not have permission to access the repo.
      `
    }

    const commits = await this.api.getPullRequestCommits()
    const comments = await this.api.getPullRequestComments()
    const activities = await this.api.getPullRequestActivities()
    const issues = await this.api.getIssues()

    return {
      metadata: this.api.repoMetadata,
      pr,
      commits,
      comments,
      activities,
      issues,
    }
  }

  supportsCommenting() {
    return true
  }

  supportsInlineComments() {
    return true
  }

  /**
   * Returns the response for the new comment
   *
   * @param {string} comment you want to post
   * @returns {Promise<any>} JSON response of new comment
   */
  createComment = (comment: string) => this.api.postPRComment(comment)

  /**
   * Makes an inline comment if possible. If platform can't make an inline comment with given arguments,
   * it returns a promise rejection. (e.g. platform doesn't support inline comments or line was out of diff).
   *
   * @returns {Promise<any>} JSON response of new comment
   */
  createInlineComment = (git: GitDSL, comment: string, path: string, line: number): Promise<any> => {
    if (!this.supportsInlineComments) {
      return new Promise((_resolve, reject) => reject())
    }
    return this.findTypeOfLine(git, line, path).then(type => {
      return this.api.postInlinePRComment(comment, line, type, path)
    })
  }

  /**
   * Finds type of line in given diff. This is needed for Bitbucket Server API
   *
   * @returns {Promise<string>} A string with type of line
   */
  findTypeOfLine = (git: GitDSL, line: number, path: string): Promise<string> => {
    console.log("\n\n\n inline ---> Finding position for inline comment." + path + "#" + line)
    return git.structuredDiffForFile(path).then(diff => {
      return new Promise<string>((resolve, reject) => {
        if (diff === undefined) {
          this.d("Diff not found for inline comment." + path + "#" + line + ". Diff: " + JSON.stringify(diff))
          reject()
        }
        this.d(
          "Diff found for inline comment, now getting a position." +
            path +
            "#" +
            line +
            ". Diff: " +
            JSON.stringify(diff)
        )
        let change
        for (let chunk of diff!.chunks) {
          // Search for a change (that is not a deletion) and with given line. We want to look only for destination lines of a change
          change = chunk.changes.find((c: any) => c.type != "del" && c.destinationLine == line)
          break
        }
        this.d("Type found for inline comment: " + JSON.stringify(change) + "." + path + "#" + line)
        resolve(change.type)
      })
    })
  }

  /**
   * Updates an inline comment if possible. If platform can't update an inline comment,
   * it returns a promise rejection. (e.g. platform doesn't support inline comments or line was out of diff).
   *
   * @returns {Promise<any>} JSON response of new comment
   */
  updateInlineComment = async (comment: string, commentId: string): Promise<any> => {
    if (!this.supportsInlineComments) {
      return new Promise((_resolve, reject) => reject())
    }
    const activities = await this.api.getPullRequestComments()
    const updateComment = activities
      .filter(activity => activity.commentAnchor)
      .map(activity => activity.comment)
      .filter(Boolean)
      .find(comment => comment!.id.toString() == commentId)

    this.d(
      "\n\n\n inline ---> Updating inline comment. CommentId: " +
        JSON.stringify(commentId) +
        "updateComment: " +
        JSON.stringify(updateComment)
    )

    return this.api.updateComment(updateComment!, comment)
  }

  /**
   * Deletes an inline comment, used when you have
   * fixed all your failures.
   *
   * @returns {Promise<boolean>} did it work?
   */
  deleteInlineComment = async (id: string): Promise<any> => {
    if (!this.supportsInlineComments) {
      return new Promise<boolean>((_resolve, reject) => reject())
    }
    const activities = await this.api.getPullRequestComments()
    const deleteComment = activities
      .filter(activity => activity.commentAnchor)
      .map(activity => activity.comment)
      .filter(Boolean)
      .find(comment => comment!.id.toString() == id)

    this.d(
      "\n\n\n inline ---> deleting inline comment. CommentId: " +
        JSON.stringify(id) +
        "deletecomment: " +
        JSON.stringify(deleteComment)
    )

    return this.api.deleteComment(deleteComment!)
  }

  /**
   * Deletes the main Danger comment, used when you have
   * fixed all your failures.
   *
   * @returns {Promise<boolean>} did it work?
   */
  deleteMainComment = async (dangerID: string): Promise<boolean> => {
    const comments = await this.api.getDangerComments(dangerID)
    for (let comment of comments) {
      await this.api.deleteComment(comment)
    }

    return comments.length > 0
  }

  /**
   * Either updates an existing comment, or makes a new one
   *
   * @param {string} newComment string value of comment
   * @returns {Promise<boolean>} success of posting comment
   */
  async updateOrCreateComment(dangerID: string, newComment: string): Promise<boolean> {
    const comments = await this.api.getDangerComments(dangerID)

    if (comments.length) {
      // Edit the first comment
      await this.api.updateComment(comments[0], newComment)

      // Delete any dupes
      for (let comment of comments) {
        if (comment !== comments[0]) {
          await this.api.deleteComment(comment)
        }
      }
    } else {
      await this.createComment(newComment)
    }

    return true
  }

  getFileContents = this.api.getFileContents
}
