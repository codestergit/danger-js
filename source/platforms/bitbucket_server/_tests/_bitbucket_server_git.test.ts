import { BitBucketServer } from "../../BitBucketServer"
import { BitBucketServerAPI } from "../BitBucketServerAPI"

import { GitCommit } from "../../../dsl/Commit"
import { FakeCI } from "../../../ci_source/providers/Fake"
import { readFileSync } from "fs"
import { resolve, join as pathJoin } from "path"
import { EOL } from "os"
import { bitBucketServerGitDSL as gitJSONToGitDSL } from "../BitBucketServerGit"

import { BitBucketServerDSL } from "../../../dsl/BitBucketServerDSL"
import { GitDSL, GitJSONDSL } from "../../../dsl/GitDSL"

const fixtures = resolve(__dirname, "..", "..", "_tests", "fixtures")

/** Returns a fixture. */
const loadFixture = (path: string): string =>
  readFileSync(pathJoin(fixtures, path), {})
    .toString()
    .replace(/\r/g, "")

/** Returns JSON from the fixtured dir */
export const requestWithFixturedJSON = async (path: string): Promise<() => Promise<any>> => () =>
  Promise.resolve(JSON.parse(loadFixture(path)))

/** Returns arbitrary text value from a request */
export const requestWithFixturedContent = async (path: string): Promise<() => Promise<string>> => () =>
  Promise.resolve(loadFixture(path))

/**
 * HACKish: Jest on Windows seems to include some additional
 * whitespace that differs from non-Windows snapshot output,
 * so we strip these until we can better-address the issue.
 */
const stripWhitespaceForSnapshot = (str: string) => {
  return str.replace(/\s/g, "")
}

const pullRequestInfoFilename = "bitbucket_server_pr.json"
const masterSHA = JSON.parse(readFileSync(pathJoin(fixtures, pullRequestInfoFilename), {}).toString()).toRef
  .latestCommit

describe("the dangerfile gitDSL - BitBucket Server", async () => {
  let bbs: BitBucketServer = {} as any
  let gitJSONDSL: GitJSONDSL = {} as any

  let bbsDSL: BitBucketServerDSL = {} as any
  let gitDSL: GitDSL = {} as any

  beforeEach(async () => {
    const api = new BitBucketServerAPI(new FakeCI({}), { host: "fake://host" })
    bbs = new BitBucketServer(api)

    api.getIssues = await requestWithFixturedJSON("bitbucket_server_issues.json")
    api.getPullRequestDiff = await requestWithFixturedJSON("bitbucket_server_diff.json")
    api.getStructuredDiff = await requestWithFixturedJSON("bitbucket_server_diff.json")
    api.getPullRequestInfo = await requestWithFixturedJSON(pullRequestInfoFilename)
    api.getPullRequestCommits = await requestWithFixturedJSON("bitbucket_server_commits.json")
    api.getPullRequestComments = await requestWithFixturedJSON("bitbucket_server_comments.json")
    api.getPullRequestActivities = await requestWithFixturedJSON("bitbucket_server_activities.json")
    api.getFileContents = async (path, repoSlug, ref) => JSON.stringify({ path, repoSlug, ref })

    gitJSONDSL = await bbs.getPlatformGitRepresentation()
    bbsDSL = await bbs.getPlatformDSLRepresentation()
    gitDSL = gitJSONToGitDSL(bbsDSL, gitJSONDSL, bbs.api)
  })

  it("sets the modified/created/deleted", async () => {
    expect(gitJSONDSL.modified_files).toEqual([".gitignore"])
    expect(gitJSONDSL.created_files).toEqual(["banana"])
    expect(gitJSONDSL.deleted_files).toEqual(["jest.eslint.config.js"])
  })

  it("shows the diff for a specific file", async () => {
    const { diff } = await gitDSL.diffForFile(".gitignore")

    expect(stripWhitespaceForSnapshot(diff)).toMatchSnapshot()
  })

  it("should include `before` text content of the file", async () => {
    const { before } = await gitDSL.diffForFile(".gitignore")

    expect(stripWhitespaceForSnapshot(before)).toMatchSnapshot()
  })

  it("should include `after` text content of the file", async () => {
    const { after } = await gitDSL.diffForFile(".gitignore")

    expect(stripWhitespaceForSnapshot(after)).toMatchSnapshot()
  })

  it("should include `added` text content of the file", async () => {
    const { added } = await gitDSL.diffForFile(".gitignore")

    expect(stripWhitespaceForSnapshot(added)).toMatchSnapshot()
  })

  it("should not include `removed` text content of the file", async () => {
    const { removed } = await gitDSL.diffForFile(".gitignore")
    expect(stripWhitespaceForSnapshot(removed)).toMatchSnapshot()
  })

  it("resolves to `null` for files not in modified_files", async () => {
    const result = await gitDSL.diffForFile("fuhqmahgads.json")

    expect(result).toBeNull()
  })

  it("sets up commit data correctly", async () => {
    expect(gitDSL.commits[0]).toMatchSnapshot()
  })

  it("shows the structured diff for a specific file", async () => {
    const { chunks } = await gitDSL.structuredDiffForFile(".gitignore")
    expect(chunks).toMatchSnapshot()
  })

  it("should have `normal` type of line for inline comment for modified file", async () => {
    const type = await bbs.findTypeOfLine(gitDSL, 3, ".gitignore")
    expect(type).toBe("normal")
  })

  it("should have `add` type of line for inline comment for modified file", async () => {
    const type = await bbs.findTypeOfLine(gitDSL, 10, ".gitignore")
    expect(type).toBe("add")
  })

  it("should have `add` type of line for inline comment for added file", async () => {
    const type = await bbs.findTypeOfLine(gitDSL, 1, "banana")
    expect(type).toBe("add")
  })

  it("checks promise rejection for line not in the diff for inline comment", async () => {
    const promise = bbs.findTypeOfLine(gitDSL, 2, "banana")
    await expect(promise).rejects
  })

  it("checks promise rejection for `del` line for inline comment for deleted file", async () => {
    const promise = bbs.findTypeOfLine(gitDSL, 0, "jest.eslint.config.js")
    await expect(promise).rejects
  })
})
