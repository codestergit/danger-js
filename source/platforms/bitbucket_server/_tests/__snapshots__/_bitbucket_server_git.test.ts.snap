// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the dangerfile gitDSL - BitBucket Server sets up commit data correctly 1`] = `
Object {
  "author": Object {
    "date": "2018-02-24T03:19:01.000Z",
    "email": "foo@bar.com",
    "name": "danger",
  },
  "committer": Object {
    "date": "2018-02-24T03:19:01.000Z",
    "email": "foo@bar.com",
    "name": "danger",
  },
  "message": "Modify and remove files",
  "parents": Array [
    "c62ada76533a2de045d4c6062988ba84df140729",
  ],
  "sha": "d6725486c38d46a33e76f622cf24b9a388c8d13d",
  "tree": null,
  "url": "fake://host/artsy/emission/commits/d6725486c38d46a33e76f622cf24b9a388c8d13d",
}
`;

exports[`the dangerfile gitDSL - BitBucket Server should include \`added\` text content of the file 1`] = `""`;

exports[`the dangerfile gitDSL - BitBucket Server should include \`after\` text content of the file 1`] = `"{\\"path\\":\\".gitignore\\",\\"repoSlug\\":\\"projects/PROJ/repos/repo\\",\\"ref\\":\\"8942a1f75e4c95df836f19ef681d20a87da2ee20\\"}"`;

exports[`the dangerfile gitDSL - BitBucket Server should include \`before\` text content of the file 1`] = `"{\\"path\\":\\".gitignore\\",\\"repoSlug\\":\\"projects/PROJ/repos/repo\\",\\"ref\\":\\"d6725486c38d46a33e76f622cf24b9a388c8d13d\\"}"`;

exports[`the dangerfile gitDSL - BitBucket Server should not include \`removed\` text content of the file 1`] = `""`;

exports[`the dangerfile gitDSL - BitBucket Server shows the diff for a specific file 1`] = `"node_modules*.log/test.py/.vscode.DS_Storecoverage.idea__pycache__/*.pyc*.swp"`;

exports[`the dangerfile gitDSL - BitBucket Server shows the structured diff for a specific file 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "content": "node_modules",
        "destinationLine": 1,
        "sourceLine": 1,
        "type": "normal",
      },
      Object {
        "content": "*.log",
        "destinationLine": 2,
        "sourceLine": 2,
        "type": "normal",
      },
      Object {
        "content": "/test.py",
        "destinationLine": 3,
        "sourceLine": 3,
        "type": "normal",
      },
      Object {
        "content": "/.vscode",
        "destinationLine": 4,
        "sourceLine": 4,
        "type": "normal",
      },
      Object {
        "content": ".DS_Store",
        "destinationLine": 5,
        "sourceLine": 5,
        "type": "normal",
      },
      Object {
        "content": "coverage",
        "destinationLine": 6,
        "sourceLine": 6,
        "type": "normal",
      },
      Object {
        "content": ".idea",
        "destinationLine": 7,
        "sourceLine": 7,
        "type": "normal",
      },
      Object {
        "content": "__pycache__/",
        "destinationLine": 8,
        "sourceLine": 8,
        "type": "normal",
      },
      Object {
        "content": "*.pyc",
        "destinationLine": 9,
        "sourceLine": 9,
        "type": "normal",
      },
      Object {
        "content": "",
        "destinationLine": 10,
        "sourceLine": 10,
        "type": "add",
      },
      Object {
        "content": "*.swp",
        "destinationLine": 11,
        "sourceLine": 10,
        "type": "normal",
      },
    ],
  },
]
`;
